core/shared/src/main/scala/zio/internal/macros/Graph.scala

```diff
@@ -9,16 +9,20 @@ final case class Graph[Key, A](
   envKeys: List[Key]
 ) {
 
-  // Map assigning to each type the times that it must be built
-  // -1 designs a `Key` from the environment
-  private var neededKeys: Map[Key, Int] = Map.empty
+  // List with the types that must be built from given nodes, contained the number of times needed
+  private var neededKeys: List[Key] = List.empty
   // Dependencies to pass to next iteration of buildComplete
   private var dependencies: List[Key]    = Nil
   private var envDependencies: List[Key] = Nil
 
+  private def get(k: Key, m: List[Key]): Int =
+    m.count(key => areEquals(key, k))
+
+  private def areEquals(k1: Key, k2: Key): Boolean =
+    keyEquals(k1, k2)
```

Comment:
Doesn't this needs to be `keyEquals(k1, k2) || keyEquals(k2, k1)`?

Reply:
Should probably have a test here to catch the "wrong" implementation.